ide/db/src/org/netbeans/modules/db/explorer/action/ConnectAction.java

```diff
@@ -62,63 +63,63 @@
 import org.openide.awt.ActionReference;
 import org.openide.awt.ActionRegistration;
 import org.openide.awt.ActionState;
-import org.openide.awt.StatusDisplayer;
+import org.openide.util.ContextAwareAction;
 import org.openide.util.HelpCtx;
 import org.openide.util.Lookup;
 import org.openide.util.Mutex;
 import org.openide.util.NbBundle;
+import org.openide.util.RequestProcessor;
+import org.openide.util.WeakListeners;
 
 @ActionRegistration(
         displayName = "#Connect", 
-        lazy = false,
+        lazy = true,
+        asynchronous = true,
```

Comment:
A pet hate pattern of mine - IMO actions should always be invoked on the EDT and then explicitly async when needed.

Are there implications to changing the behaviour of an existing action here?

Reply:
Yes. In NB IDE it does only a little harm, as waiting on the dialog creates a nested event loop, and everything incl the 'blocked' EDT thread works (as the nested event loop services events). 
But in headless mode, if the [action's code blocks](https://github.com/apache/netbeans/pull/5361/files#diff-033fc81193b23efa97939feead127364383d694ff672ba7d5d1612c37cf499dfL331), the caller message handling thread blocks, so the LSP client gets no progress messages etc.

There were two options, actually: explicit offload the work to a RequestProcessor if headless, or this one. I'd slightly prefer `asynchronous = true` as it allows (eventually) the infrastructure to track what actions are running (not used now).

Reply:
Generally prefer the RequestProcessor approach, but hadn't considered the infrastructure tracking side.  Just been caught a few times (mainly in RCP work) by actions triggering chains of things unexpectedly off of the EDT.

I understand why you want to change it for the headless mode.  I assume there are no existing usages where the action not being completed by return is a problem?

Anyway, assuming no implications to changing the existing action, no concerns with it as it is.